arroyo-worker/src/connectors/kinesis/source/mod.rs

```diff
@@ -0,0 +1,552 @@
+                            }
+                            retries += 1;
+                            tokio::time::sleep(Duration::from_millis(200 * (1 << retries))).await;
+                        } else {
+                            return Err(anyhow!(error));
+                        }
+                    }
+                    _ => return Err(anyhow!(error)),
+                },
+            }
+        }
+    }
+
+    async fn handle_get_records(
+        &mut self,
+        shard_id: String,
+        get_records: GetRecordsOutput,
+        ctx: &mut Context<(), T>,
+    ) -> Result<Option<BoxedFuture<AsyncNamedResult<AsyncResult>>>, UserError> {
+        let last_sequence_number = get_records.records().and_then(|records| {
+            records
+                .last()
+                .map(|record| record.sequence_number().unwrap().to_owned())
+        });
+
+        let next_shard_iterator = self.process_records(get_records, ctx).await?;
+        let shard_state = self.shards.get_mut(&shard_id).unwrap();
+
+        if let Some(last_sequence_number) = last_sequence_number {
+            shard_state.offset = KinesisOffset::SequenceNumber(last_sequence_number.to_string());
+        }
+
+        match next_shard_iterator {
+            Some(shard_iterator_id) => Ok(Some(self.next_read_future(shard_id, shard_iterator_id))),
+            None => {
+                shard_state.closed = true;
+                Ok(None)
+            }
+        }
+    }
+    async fn handle_need_new_iterator(
+        &mut self,
+        shard_id: String,
+    ) -> Result<Option<BoxedFuture<AsyncNamedResult<AsyncResult>>>, UserError> {
+        let shard_state = self.shards.get_mut(&shard_id).unwrap();
+        Ok(Some(shard_state.get_update_shard_iterator_future(
+            self.kinesis_client.as_ref().unwrap(),
+        )))
+    }
+
+    fn get_shard_iterator_call(&self) -> GetShardIterator {
+        self.kinesis_client.as_ref().unwrap().get_shard_iterator()
+    }
+
+    /// Runs the Kinesis source, handling incoming records and control messages.
+    ///
+    /// This method initializes the Kinesis client, initializes the shards, and enters a loop to handle incoming
+    /// records and control messages. There are three prongs to the tokio select loop:
+    /// * A `FuturesUnordered` tha contains futures for reading off of shards.
+    /// * An interval that periodically polls for new shards, initializing their futures.
+    /// * Polling off of the control queue, to perform checkpointing and stop the operator.
+    async fn run_int(&mut self, ctx: &mut Context<(), T>) -> Result<SourceFinishType, UserError> {
+        let config = load_from_env().await;
+        self.kinesis_client = Some(KinesisClient::new(&config));
+        let starting_futures = self.init_shards(ctx).await.map_err(|e| {
+            UserError::new(
+                "failed to initialize shards. Underlying error:",
```

Comment:
The first parameter of the user error should be an error name; it's not concatenated with the error details. So this could just be "failed to initialize shards"